hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/common/table/read/TestIfRecordPositionColumnWorkForSpark3_5.scala

```diff
@@ -0,0 +1,105 @@
+ * under the License.
+ */
+
+package org.apache.hudi.common.table.read
+
+import org.apache.hadoop.conf.Configuration
+import org.apache.hudi.SparkAdapterSupport.sparkAdapter
+import org.apache.hudi.common.config.HoodieStorageConfig
+import org.apache.hudi.common.testutils.HoodieTestTable
+import org.apache.hudi.config.HoodieWriteConfig
+import org.apache.hudi.testutils.SparkClientFunctionalTestHarness
+import org.apache.hudi.util.CloseableInternalRowIterator
+import org.apache.hudi.{DataSourceWriteOptions, HoodieSparkUtils}
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.execution.datasources.parquet.{HoodieFileGroupReaderBasedParquetFileFormat, ParquetFileFormat}
+import org.apache.spark.sql.types.{IntegerType, StringType, StructField, StructType}
+import org.junit.jupiter.api.Assertions.{assertEquals, assertTrue}
+import org.junit.jupiter.api.Test
+
+class TestIfRecordPositionColumnWorkForSpark3_5 extends SparkClientFunctionalTestHarness {
+  private val HOOD_FORMAT = "org.apache.hudi"
+  private val MOR_TABLE_TYPE = "MERGE_ON_READ"
+  private val PARQUET_FORMAT = "parquet"
+  private val ROW_INDEX_COLUMN = "_tmp_metadata_row_index"
+  private val SPARK_MERGER = "org.apache.hudi.HoodieSparkRecordMerger"
+
+  @Test
+  def testConfirmIfRowIndexColumnWorks(): Unit = {
+    val _spark = spark
+    import _spark.implicits._
+
+    val userToCountryDF = Seq(
+      (1, "US", "1000"),
+      (2, "US", "1000"),
+      (3, "China", "1000"),
+      (4, "Singapore", "1000")
+    ).toDF("userid", "country", "ts")
+    val userToCountryTblPath = s"$basePath/user_to_country_table"
+
+    userToCountryDF.write.format(HOOD_FORMAT)
+      .option(DataSourceWriteOptions.RECORDKEY_FIELD.key, "userid")
+      .option(DataSourceWriteOptions.PRECOMBINE_FIELD.key, "ts")
+      .option(HoodieWriteConfig.TBL_NAME.key, "user_to_country")
+      .option(HoodieWriteConfig.RECORD_MERGER_IMPLS.key, SPARK_MERGER)
+      .option(HoodieWriteConfig.WRITE_RECORD_POSITIONS.key, "true")
+      .option(HoodieStorageConfig.LOGFILE_DATA_BLOCK_FORMAT.key, PARQUET_FORMAT)
+      .option(DataSourceWriteOptions.TABLE_TYPE.key(), MOR_TABLE_TYPE)
+      .save(userToCountryTblPath)
+
+    val dataSchema = new StructType(
+      Array(
+        StructField("userid", IntegerType, nullable = false),
+        StructField("country", StringType, nullable = false),
+        StructField("ts", StringType, nullable = false)
+      )
+    )
+    val requiredSchema = HoodieFileGroupReaderBasedParquetFileFormat.getAppliedRequiredSchema(
+      dataSchema, shouldUseRecordPosition = true, ROW_INDEX_COLUMN)
+    if (HoodieSparkUtils.gteqSpark3_5) {
+      assertEquals(4, requiredSchema.fields.length)
+    }
+
+    _spark.conf.set("spark.sql.parquet.enableVectorizedReader", "false")
+    val metaClient = getHoodieMetaClient(
+      spark().sparkContext.hadoopConfiguration, userToCountryTblPath)
+    val fileReader = new ParquetFileFormat().buildReaderWithPartitionValues(
```

Comment:
Could you directly test `preMergeBaseFileReader` returned from `HoodieFileGroupReaderBasedParquetFileFormat#buildFileReaders` and validate the row positions returned from the Spark parquet reader?

Reply:
> Could you directly test `preMergeBaseFileReader` returned from `HoodieFileGroupReaderBasedParquetFileFormat#buildFileReaders` and validate the row positions returned from the Spark parquet reader?

Exposing preMergeBaseFileReader would take some extra efforts, and did not help much since we construct them from the same function.